functions.lisp

```diff
@@ -159,55 +191,90 @@ For generic implementation and getails, see `jrem'."
       (maphash (lambda (key val)
                  (setf (gethash key new) val))
                object)
       new))
-  (:documentation "Copy the OBJECT, potentially creating an identical one."))
-
-(defun copy_ (object)
-  "Copy the OBJECT, potentially creating an identical one.
-
-For generic implementation and getails, see `jcopy'."
-  (warn 'deprecated :deprecated 'copy_ :replacement "NJSON/ALIASES:COPY")
-  (jcopy object))
+  (:documentation "Copy the OBJECT, potentially creating an identical one.
+Coerce all JSON arrays to adjustable vectors."))
+
+(defgeneric jkeys (object)
+  (:method ((object vector))
+    (loop for i from 0 below (length object)
+          collect i))
+  (:method ((object string))
+    (cerror "Return nothing"
+            'non-indexable :value object))
+  (:method ((object hash-table))
+    (loop for key being the hash-key of object
+          collect key))
+  (:method ((object t))
+    (cerror "Return nothing"
+            'non-indexable :value object))
+  (:documentation "Get keys to index OBJECT with, as a list of integers/strings.
+If the OBJECT is not a JSON array/object, throws `non-indexable'."))
 
 (defgeneric jtruep (object)
   (:method (object)
     (declare (ignore object))
     t)
-  ;; TODO: A method on numbers, checking for zero?
-  (:method ((object sequence))
+  ;; Questionable: empty strings are false.
+  (:method ((object string))
     (not (uiop:emptyp object)))
-  (:method ((object hash-table))
-    (plusp (hash-table-count object)))
+  ;; Questionable: zero numbers are false.
+  (:method ((object number))
+    (not (zerop object)))
   (:method ((object symbol))
-    (not (member object (list nil :null))))
-  (:documentation "Test OBJECT for truthiness in JSON terms."))
+    (not (member object (list nil :null :undefined))))
+  (:documentation "Test OBJECT for truthiness in JavaScript terms."))
 
 (dolist (symbol '(jtrue-p jtrue?))
   (setf (symbol-function symbol) #'jtruep))
 
-(macrolet ((defalias (name)
-             `(defun ,name (object)
-                "Test OBJECT for truthiness in JSON terms.
-For generic implementation, see `jtruep'."
-                (warn 'deprecated :deprecated (quote ,name) :replacement "NJSON/ALIASES:TRUE")
-                (jtruep object))))
-  (defalias truep)
-  (defalias true-p)
-  (defalias true?))
-
 (defun jnot (arg)
   "JSON-aware version of `cl:not'."
   (not (jtruep arg)))
 
-(defun not_ (arg)
-  "JSON-aware version of `cl:not'.
-Alias for `jnot'."
-  (warn 'deprecated :deprecated 'not_ :replacement "NJSON/ALIASES:NOT")
-  (jnot arg))
+(defgeneric ensure-array (object &key &allow-other-keys)
+  (:method ((object hash-table) &key convert-objects &allow-other-keys)
+    (if convert-objects
+        (make-array (hash-table-count object)
+                    :adjustable t
+                    :fill-pointer t
+                    :initial-contents (loop for key in (jkeys object)
+                                            collect (jget key object)))
```

Comment:
Replace `loop` with `mapcar`?

Reply:
This `mapcar` will require closing over `object`, which would be hard without `alex:curry` (because Alexandria is not a dependency). So `loop` it is then.